ob-swiftui.el

```diff
@@ -109,23 +109,35 @@
 ;; Aliasing enables block syntax highlighting.
 (defalias 'swiftui-mode #'swift-mode)
 
-(defvar org-babel-default-header-args:swiftui '((:results . "window")
-                                                (:view . "none"))
+(defvar org-babel-default-header-args:swiftui
+  '((:results . "window")
+    (:view . "none")
+    (:file . (lambda ()
+               (make-temp-file "ob-swiftui-output-file-" nil ".png"))))
   "Default ob-swiftui header args.
 Must be named `org-babel-default-header-args:swiftui' to integrate with `ob'.")
 
 (defun org-babel-execute:swiftui (body params)
   "Execute a block of SwiftUI code in BODY with org-babel header PARAMS.
 This function is called by `org-babel-execute-src-block'"
   (message "executing SwiftUI source code block")
-  (let ((target (make-temp-file "ob-swiftui-exe-")))
+  (let* ((write-to-file (member "file" (map-elt params :result-params)))
+         (output-file (and write-to-file
+                           (let ((file (cdr (assoc :file params))))
+                             (cond
+                              ((functionp file) (funcall file))
+                              ((stringp file) file)
+                              (t nil)))))
+         (target (make-temp-file "ob-swiftui-exe-")))
     (with-temp-buffer
       (insert (ob-swiftui--expand-body body params))
       (shell-command-on-region
        (point-min)
        (point-max)
-       (format "swiftc -o %s -" target) nil))
-    (shell-command-to-string target)))
+       (format "swiftc -o %s -" target)))
+    (shell-command-to-string (format "%s > %s" target
```

Comment:
> If the SwiftUI block doesn't compile (ie. introduce any invalid code), the compilation error is not shown to the user.
Should be fixed with https://github.com/xenodium/ob-swiftui/pull/4/commits/16ad71b0c75830e987c0737471b2779c2a6180e9

>  However, if there is any other output/logs (or future runtime warnings out of our control) won't this break the generated image?
No, warnings/logs are belong to standard error, not standard output IMO. If this has problem, then the previous method which output the file path directly is on the same situation. I've considered your option, but it's hard to construct the right output file path inside Swift code. However maybe it's due to my knowledge limitation, if you have any idea please help.

Reply:
> No, warnings/logs are belong to standard error, not standard output IMO

Indeed, but we don't have enough control to enforce it. If SwiftUI implementation (or any dependency) outputs any messages to standard output, it would start corrupting the file written.

> then the previous method which output the file path directly is on the same situation(not meaning image generation, but file path may be mess-up with other outputs as you described)

There may be noise output before or after the generated path but the path itself should not be affected. Worse case scenario, the user manually opens the file but the file is never corrupt.

> if you have any idea please help.

Not tested this, but I'm thinking something like the following:

```
diff --git a/ob-swiftui.el b/ob-swiftui.el
index cf39d0e..31b9556 100644
--- a/ob-swiftui.el
+++ b/ob-swiftui.el
@@ -151,7 +151,7 @@ This function is called by `org-babel-execute-src-block'"
       (message (ob-swiftui--expand-body-window body root-view))
       (ob-swiftui--expand-body-window body root-view))))
 
-(defun ob-swiftui--expand-body-preview (body root-view)
+(defun ob-swiftui--expand-body-preview (body root-view output-file)
   (format
      "
 import SwiftUI
@@ -162,8 +162,7 @@ let timer = Timer.scheduledTimer(withTimeInterval: 0.1, repeats: false) { timer
     renderer.scale = NSScreen.main?.backingScaleFactor ?? 1.0
     let data = await renderer.cgImage?.pngData(compressionFactor: 1)
     do {
-      let url = URL(fileURLWithPath: NSTemporaryDirectory(), isDirectory: true)
-        .appendingPathComponent(ProcessInfo.processInfo.globallyUniqueString + \".png\")
+      let url = URL(fileURLWithPath: \"%s\")
       try data?.write(to: url)
       print(url.path, terminator: \"\")
       exit(0)
@@ -198,7 +197,8 @@ struct ContentView: View {
     .frame(maxWidth:.infinity, maxHeight:.infinity)
   }
 }
-" body))))
+" body))
+     output-file))
```

Hope this makes sense.

Reply:
> Should be fixed with https://github.com/xenodium/ob-swiftui/commit/16ad71b0c75830e987c0737471b2779c2a6180e9

Thank you!

Reply:
> Hope this makes sense.

You are right. I don't know why I stuck before. I've change the way. Note, output the path is not necessary, and we don't rely the output at all from now on.